### ISSUE TITLE
[CI] Build and publish container images

### ISSUE THREAD

**Issue by okdas** [pokt-network/poktroll#106]
## Objective

We need to build container images so that they can be utilized first by us on dev- and test-nets, and later by the community.

## Origin Document

This became necessary as we worked on DevNets. We need container images to run DevNets and execute various types of tests on them.

## Goals

- Build containers so that they can be utilized for tests.

## Deliverables

- [x] Container images hooked to GHCR.io
- [x] DevNets utilizing these images

## Non-goals / Non-deliverables

- Build production-ready, slim container images

## General deliverables

- [ ] **Comments**: Add/update TODOs and comments alongside the source code so it is easier to follow.
- [ ] **Testing**: Add new tests (unit and/or E2E) to the test suite.
- [ ] **Makefile**: Add new targets to the Makefile to make the new functionality easier to use.
- [ ] **Documentation**: Update architectural or development READMEs; use [mermaid](https://mermaid-js.github.io/mermaid/) diagrams where appropriate.

---

**Creator**: @okdas

**Comment by Olshansk** [comment]
@okdas Couple questions about this:

1. Are we going with https://ghcr.io/ because all of our other dev is on GitHub? I like the approach but just wanted to understand why.

2. Will these be built on merges to `main` or did you have something else in mind?

3. Will we have custom containers for each actor (app, supplier, gateway, etc...)?

**Comment by okdas** [comment]
@Olshansk  
1. The reason we switched from Docker Hub is due to lost access to one of the accounts. I have continued to use GHCR.  
Pros:  
- Free  
- Well integrated into GitHub UI  
- Unlike Docker Hub, GHCR rate limiting is not severe. On Docker Hub, anonymous pulls are [easy to exhaust](https://docs.docker.com/docker-hub/download-rate-limit/).  
Cons:  
- Unlike Docker Hub, the image names must include `ghcr.io`. For example, `ghcr.io/pokt-network/pocketd`, whereas on Docker Hub, we could simply use `pokt-network/pocketd`.  

2. We need images built on non-main branches for DevNets. We will also build on the main branch because why not - let's make sure they can be built after the merge.  
3. Not until we separate the binary into actor-specific components.
